src/executor.ts

```diff
@@ -338,86 +253,94 @@ export class TaskExecutor {
    * @private
    */
   private async doShutdown() {
-    this.events.emit("beforeEnd", Date.now());
+    this.events.emit("executorBeforeEnd", Date.now());
     if (isNode) this.removeSignalHandlers();
     clearTimeout(this.startupTimeoutId);
-    if (!this.configOptions.storageProvider) await this.storageProvider?.close();
-    await this.networkService?.end();
-    await Promise.all([this.taskService.end(), this.agreementPoolService.end(), this.marketService.end()]);
-    await this.paymentService.end();
-    await this.yagna.end();
-    // this.options.eventTarget?.dispatchEvent(new Events.ComputationFinished());
+    await this.taskService?.end();
+    await this.resourceRentalPool?.drainAndClear();
+    await this.glm.disconnect();
     this.printStats();
     await this.statsService.end();
     this.logger.info("Task Executor has shut down");
-    this.events.emit("end", Date.now());
+    this.events.emit("executorEnd", Date.now());
   }
 
   getStats() {
     return {
       ...this.statsService.getAll(),
-      retries: this.taskService.getTotalRetryCount(),
+      retries: this.taskService?.getTotalRetryCount(),
     };
   }
 
   /**
-   * Registers a worker function that will be run when an activity is ready.
+   * Registers a setup function that will be run when an exe-unit is ready.
    * This is the perfect place to run setup functions that need to be run only once per
-   * activity, for example uploading files that will be used by all tasks in the activity.
-   * This function can be called multiple times, each worker will be run in the order
-   * they were registered.
+   * exe-unit, for example uploading files that will be used by all tasks in the exe-unit.
+   * This function can be called multiple times.
    *
-   * @param worker worker function that will be run when an activity is ready
+   * @param setup setup function that will be run when an exe-unit is ready
    * @example
    * ```ts
-   * const uploadFile1 = async (ctx) => ctx.uploadFile("./file1.txt", "/file1.txt");
-   * const uploadFile2 = async (ctx) => ctx.uploadFile("./file2.txt", "/file2.txt");
+   * const uploadFile = async (exe) => exe.uploadFile("./file1.txt", "/file1.txt");
    *
-   * executor.onActivityReady(uploadFile1);
-   * executor.onActivityReady(uploadFile2);
+   * executor.onExeUnitReady(uploadFile);
    *
-   * await executor.run(async (ctx) => {
-   *  await ctx.run("cat /file1.txt /file2.txt");
+   * await executor.run(async (exe) => {
+   *  await exe.run("cat /file1.txt /file2.txt");
    * });
    * ```
    */
-  onActivityReady(worker: Worker<unknown>) {
-    this.activityReadySetupFunctions.push(worker);
+  onExeUnitReady(setup: LifecycleFunction) {
+    this.setupFunction = setup;
```

Comment:
you are right, but I suggested consistency with our new API, there we have options called `setup` and `teardown` and here it is similar... We can easily change it according to your suggestion, but the parameter names would have to be changed to `onExeUnitReady(setupFunctions: LifecycelFunction[])` so I had the feeling that `setup: LifecycleFunction` would be simpler and sufficient, especially since you can always perform such a sequence inside your setup functon and I don't see a special value in the array

Reply:
I was looking at the old tickets that introduced this change: https://golemproject.atlassian.net/browse/JST-101 and I'm not quite sure why we allowed multiple setup functions in the first place? I'm ok with dropping this and going back to only one, but I'd like @grisha87 to sign off on that change too.

Reply:
Having a single setup function for a rental is just enough. It's harder to make an overkill by the user. In context of the TE, I think this also can be a single one, but then the prefix `onX` doesn't fit as it gives the notion of attaching one or more handlers for the event.

> why we allowed multiple setup functions in the first place?

I think that was supposed to open possibilities, but judging it with the current experience in using the SDK in mind: we don't need to have many setup functions in the SDK. No plugin use-case comes in mind either. Setup functions are tied to specific types of images or workloads so it makes sense to define them together in the same place. Just like we do it with `oneOf` or `manyOf`.

Now, if I were to translate where the setup function would need to be in TE - then I would suggest making it available as part of `TaskExecutor.create({ setup: () => {}, teardown: () => {})` and then just drop the `onExeUnitReady` altogether.

Reply:
yeah I agree, passing setup and teardown in task executor options makes the most sense, I think the right place for this is inside `task` options:
```ts
const executor = await TaskExecutor.create({
  demand: {
    workload: {
      imageTag: "golem/alpine:latest",
    },
  },
  market: {
    rentHours: 0.5,
    pricing: {
      model: "linear",
      maxStartPrice: 0.5,
      maxCpuPerHourPrice: 1.0,
      maxEnvPerHourPrice: 0.5,
    },
  },
  task: {
    maxParallelTasks: 5,
    setup: () => {...},
    teardown: () => {...},
  }
});
```

Reply:
For me too, this is a better solution 👍